File: php/src/Advertisements/Advertisement/UI/Http/DeleteAdvertisementController.php
<?php
declare(strict_types=1);

namespace Demo\App\Advertisements\Advertisement\UI\Http;

use Demo\App\Advertisements\Advertisement\Application\Command\DeleteAdvertisement\DeleteAdvertisementCommand;
use Demo\App\Advertisements\Advertisement\Application\Command\DeleteAdvertisement\DeleteAdvertisementUseCase;
use Demo\App\Common\Exceptions\BoundedContextException;
use Demo\App\Common\UI\CommonController;
use Demo\App\Framework\FrameworkRequest;
use Demo\App\Framework\FrameworkResponse;
use Demo\App\Framework\SecurityUser\FrameworkSecurityService;

final class DeleteAdvertisementController extends CommonController
{
    public function __construct(
        private DeleteAdvertisementUseCase $useCase,
        private FrameworkSecurityService $securityService,
    ) {}

    public function request(FrameworkRequest $request, array $pathValues = []): FrameworkResponse
    {
        try {
            $user = $this->securityService->getSecurityUserFromRequest($request);

            if (null == $user || !$user->role() == 'member') {
                return $this->processUnauthorizedResponse();
            }

            $command = new DeleteAdvertisementCommand(
                $user->id(),
                $user->role(),
                $pathValues['advertisementId'],
            );

            $this->useCase->execute($command);

            return $this->processSuccessfulCommand();
        } catch (BoundedContextException $exception) {
            return $this->processDomainOrApplicationExceptionResponse($exception);
        } catch (\Throwable $exception) {
            return $this->processGenericException($exception);
        }
    }
}

File: kotlin/src/main/kotlin/advertisements/advertisement/ui/http/DeleteAdvertisementController.kt
package advertisements.advertisement.ui.http

import advertisements.advertisement.application.command.deleteadvertisement.DeleteAdvertisementCommand
import advertisements.advertisement.application.command.deleteadvertisement.DeleteAdvertisementUseCase
import common.application.ElementNotFoundException
import common.exceptions.BoundedContextException
import common.ui.http.CommonController
import framework.FrameworkRequest
import framework.FrameworkResponse
import framework.securityuser.FrameworkSecurityService

class DeleteAdvertisementController(
    private val useCase: DeleteAdvertisementUseCase,
    private val securityService: FrameworkSecurityService,
): CommonController() {

    override fun execute(request: FrameworkRequest, pathValues: Map<String, String>): FrameworkResponse {
        try {
            val user = securityService.getSecurityUserFromRequest(request)!!

            if (user.role != "member") {
                return processUnauthorizedResponse()
            }

            useCase.execute(
                DeleteAdvertisementCommand(
                    user.id,
                    user.role,
                    pathValues["advertisementId"]!!,
                )
            )

            return processSuccessfulCommand()
        } catch (e: ElementNotFoundException) {
            return processNotFoundCommand(e)
        } catch (e: BoundedContextException) {
            return processApplicationOrDomainException(e)
        } catch (e: Exception) {
            return processGenericException(e)
        }
    }
}